apps/web/src/components/WalletConnect/WalletConnectProvider.tsx

```diff
@@ -0,0 +1,119 @@
+import { type NetworkType } from "@airgap/beacon-wallet";
+import { useToast } from "@chakra-ui/react";
+import { type WalletKitTypes } from "@reown/walletkit";
+import { useDynamicModalContext } from "@umami/components";
+import {
+  createWalletKit,
+  useAsyncActionHandler,
+  useAvailableNetworks,
+  useWcPeers,
+  walletKit,
+} from "@umami/state";
+import { type Network } from "@umami/tezos";
+import { formatJsonRpcError } from "@walletconnect/jsonrpc-utils";
+import { getSdkError } from "@walletconnect/utils";
+import { type PropsWithChildren, useEffect } from "react";
+
+import { SessionProposalModal } from "./SessionProposalModal";
+
+export const context = {};
+
+export const WalletConnectProvider = ({ children }: PropsWithChildren) => {
+  const onSessionProposal = useOnSessionProposal();
+  const onSessionDelete = useOnSessionDelete();
+  const onSessionRequest = useOnSessionRequest();
+
+  useEffect(() => {
+    const initializeWallet = async () => {
+      await createWalletKit();
+      walletKit.on("session_proposal", event => void onSessionProposal(event));
+      walletKit.on("session_request", event => void onSessionRequest(event));
+      walletKit.on("session_delete", event => void onSessionDelete(event));
+    };
+
+    void initializeWallet();
+  });
+
+  return children;
+};
+
+const useOnSessionProposal = () => {
+  const { handleAsyncActionUnsafe } = useAsyncActionHandler();
+  const { openWith } = useDynamicModalContext();
+  const availableNetworks: Network[] = useAvailableNetworks();
+
+  return (proposal: WalletKitTypes.SessionProposal) =>
+    handleAsyncActionUnsafe(async () => {
+      // dApp sends in the session proposal the required networks and the optional networks.
+      // The response must contain all the required networks but Umami supports just one per request.
+      // So if the list of required networks is more than one or the required network is not supported, we can only reject the proposal.
+      const requiredNetworks = Object.entries(proposal.params.requiredNamespaces)
+        .map(([key, values]) => (key.includes(":") ? key : values.chains))
+        .flat();
+
+      if (requiredNetworks.length !== 1 || requiredNetworks[0] === undefined) {
```

Comment:
If you want to check if `there are more than 1 network or networks array is empty` you can simplify to 

`if (requiredNetworks.length > 1 || !requiredNetworks.length) {}`